src/modules/job-exec/sdexec.c

```diff
@@ -0,0 +1,584 @@
+                  LOG_ERR,
+                  "unable run multiuser job with no IMP configured!");
+        return -1;
+    }
+
+    if (!(se = sdexec_create (h, job, config_get_job_shell (job))))
+        return -1;
+
+    job->data = se;
+    return 1;
+}
+
+static void state_cb (flux_sdprocess_t *sdp, flux_sdprocess_state_t state, void *arg)
+{
+    struct sdexec *se = arg;
+    if (state == FLUX_SDPROCESS_ACTIVE) {
+        if (se->job->multiuser) {
+            char *input = NULL;
+            json_t *o = json_pack ("{s:s}", "J", se->job->J);
+            if (!o || !(input = json_dumps (o, JSON_COMPACT))) {
+                jobinfo_fatal_error (se->job, errno, "Failed to get input to IMP");
+                return;
+            }
+            if (write (se->stdin_fds[0], input, strlen (input)) < 0) {
+                jobinfo_fatal_error (se->job, errno, "write");
+                return;
+            }
+            close (se->stdin_fds[0]);
+            se->stdin_fds[0] = -1;
+        }    
+    }
+    else if (state == FLUX_SDPROCESS_EXITED
+             && !se->jobinfo_tasks_complete_called) {
+        jobinfo_tasks_complete (se->job,
+                                resource_set_ranks (se->job->R),
+                                flux_sdprocess_wait_status (se->sdp));
+        se->jobinfo_tasks_complete_called = true;
+    }
+}
+
+static void stdout_cb (flux_reactor_t *r, flux_watcher_t *w,
+                       int revents, void *arg)
+{
+    struct sdexec *se = arg;
+    const char *cmd = flux_cmd_arg (se->cmd, 0);
+    if (revents & FLUX_POLLIN) {
+        char buf[1024];
+        int len;
+        if ((len = read (se->stdout_fds[0], buf, 1024)) < 0)
+            jobinfo_fatal_error (se->job, errno, "read");
+        else if (len > 0)
+            jobinfo_log_output (se->job,
+                                0,
+                                basename (cmd),
+                                "stdout",
+                                buf,
+                                len);
+    }
+}
+
+static void stderr_cb (flux_reactor_t *r, flux_watcher_t *w,
+                       int revents, void *arg)
+{
+    struct sdexec *se = arg;
+    const char *cmd = flux_cmd_arg (se->cmd, 0);
+    if (revents & FLUX_POLLIN) {
+        char buf[1024];
+        int len;
+        if ((len = read (se->stderr_fds[0], buf, 1024)) < 0)
+            jobinfo_fatal_error (se->job, errno, "read");
+        else if (len > 0)
+            jobinfo_log_output (se->job,
+                                0,
+                                basename (cmd),
+                                "stderr",
+                                buf,
+                                len);
+    }
+}
+
+static int sdexec_start (struct jobinfo *job)
+{
+    struct sdexec *se = job->data;
+    char *unitname = NULL;
+    char **cmdv = NULL;
+    char **envv = NULL;
+    int rv = -1;
+
+    if (asprintf (&unitname, "flux-sdexec-%ju", (uintmax_t)job->id) < 0) {
```

Comment:
Something I just considered: The unitname for jobs should include the current rank as well as jobid in case we are running tests with multiple brokers per node.

Reply:
good idea!